### ISSUE TITLE
ECJ support

### ISSUE THREAD

**Issue by ascopes** [ascopes/java-compiler-testing#163]
Reconsider whether ECJ support is feasible. This was removed in #92 due to the development overhead for working around bugs in ECJ's JSR-199 compliance, but with some architectural changes in this library that have been added since that issue was closed, it may be feasible to consider supporting ECJ again.

**Comment by ascopes** [comment]
As of right now, the following issues have been found for doing this.

## Fixable stuff/stuff that I can work around

- ECJ seems to break when handling Java modules on Java 11 back compiling to Java 9 or Java 10. I can add an exclusion to prevent this running if needed.
- ECJ is passing invalid binary names around that use slash-delimiting rather than period delimiting. I can add a workaround to detect this and reformat as needed.
- ECJ is incompatible with Checkerframework. I can skip tests for this.
- ECJ will not work with RAM directories. I can use temp directories for this instead with some internal tweaking.
- ECJ flags differ to Javac for -Werror. I can just change the flag value.

## Unfixable stuff

- Compiling a single source tree with a module-info in it that is not a multiple-module source tree results in ECJ making a nested module directory for the source. Javac does not do this.
- ECJ fails to look up modules correctly in the classpath.
- ECJ diagnostics render incorrectly due to internal bugs in ECJ.
- ECJ throws a NullPointerException trying to compile Java 9 sources in Java 11 -- no idea why.
- ECJ complains when two packages overlap for modules. Javac deals with this fine.

**Comment by ascopes** [comment]
Putting this on hold as I feel like this is more work than it is worth right now.

Honestly I probably need an ECJ dev to help with this as I do not know their code base well enough to work out what the problems are here.

**Comment by ascopes** [comment]
If I do implement this, I will probably use a different Maven module to make ECJ support optional, going forwards.

Keeping the ECJ branch as there is some useful stuff on there for reference, but the actual implementation will need a new branch since `main` has diverged by 250+ commits.

**Comment by ascopes** [comment]
Blocked by https://github.com/eclipse-jdt/eclipse.jdt.core/issues/958 which prevents modules being detected for custom JavaFileManager objects

**Comment by ascopes** [comment]
Per https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1153, it appears ECJ will no longer support Java 11, meaning any changes needed to fix outstanding issues and bugs we are seeing will not be supported on Java 11 if/when they are considered.

For now, I am going to close this issue as not planned, along with the PR to implement ECJ support.

**Comment by agentgt** [comment]
@ascopes Eclipse file manager is not the only one that has bizarre behavior. Gradle incremental does unusual things as well and coincidentally slightly similar has problems with modules: https://github.com/jstachio/jstachio/issues/223 albeit not the same problem.

I have been meaning to try to use a framework such as yours on my project but because most of the bugs have been integration (Eclipse, Maven, Gradle, Intellij etc) I blackbox generate code to the filesystem (then check it in which is kind of messed up but those tests do not change much).

**Comment by ascopes** [comment]
@agentgt yeah, it is a bit of a pain point.

So far, we appear to be fully compatible with how maven-compiler-plugin exposes file management under javac.

ECJ has been a bit of a nightmare as ECJ has several fairly major bugs regarding modules if you do not use their custom file manager implementation. I've spent several hours trying to find a workaround but in the end the only solution would be maintaining a fixed fork. Some of the issue as well is that ECJ doesn't consistently use the file manager interfaces, sometimes falling back to just using java.io.File directly, which is a fairly big issue as you lose the ability to control the file system being run on.

I haven't fully tested Gradle integration, but this is definitely something I'd like to see included in this library as a test pack.

Since file management in this library uses a scratch-written implementation, we have the ability to fully control exactly how it works and tune it to whatever is needed.

If there is anything missing that you think would be useful for clearing up some of those concerns you mentioned, I am more than happy to discuss it!

**Comment by ascopes** [comment]
@agentgt just a small FYI. Eclipse have picked up the issues I've raised with them and are looking into the fixes required on their side. I've got a provisional branch up at GH-773 that plans to include ECJ in v5.0.0 of this library if we can get all tests passing. It is still not confirmed yet but just thought it was worth pinging you an update 🙂
